graylog2-server/src/main/java/org/graylog2/security/sessions/MongoDbSessionService.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.security.sessions;
+
+import com.google.errorprone.annotations.MustBeClosed;
+import com.mongodb.client.model.IndexOptions;
+import com.mongodb.client.model.Indexes;
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+import org.graylog2.database.MongoCollection;
+import org.graylog2.database.MongoCollections;
+import org.graylog2.database.utils.MongoUtils;
+import org.graylog2.events.ClusterEventBus;
+import org.graylog2.security.SessionDeletedEvent;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static com.mongodb.client.model.Filters.eq;
+import static org.graylog2.security.sessions.SessionDTO.FIELD_SESSION_ID;
+
+@Singleton
+public class MongoDbSessionService implements SessionService {
+    public static final String COLLECTION_NAME = "sessions";
+    private final ClusterEventBus eventBus;
+    private final MongoCollection<SessionDTO> collection;
+
+    @Inject
+    public MongoDbSessionService(MongoCollections mongoCollections, ClusterEventBus clusterEventBus) {
+        this.eventBus = clusterEventBus;
+        this.collection = mongoCollections.collection(COLLECTION_NAME, SessionDTO.class);
+
+        // Legacy: we had a non-unique index on session_id before. Remove it so that we can create the index again as
+        // unique.
+        final var indexes = collection.listIndexes().into(new ArrayList<>());
+        indexes.stream()
+                .filter(index -> index.getString("name").equals("session_id_1"))
+                .filter(index -> !index.getBoolean("unique", false))
+                .findFirst()
+                .ifPresent(index -> collection.dropIndex("session_id_1"));
```

Comment:
The index deletion may fail when multiple nodes execute this code simultaneously. Let's wrap it in a try-catch block.